### PR TITLE
Easier Clang Integration, Type Analysis Updates, & Primitive Cuda support

### PR DESCRIPTION
**Pull request by wsmoses**
This introduces a ClangEnzyme.so for loading directly into clang (necessary for ease of use with GPU backend). Introduces primitive support for PTX backend among various type and activity analysis updates. Also fixes pre and post Enzyme optimizations to work across various LLVM versions

### CODE REVIEW

File: enzyme/Enzyme/Enzyme.cpp [hunk 1]
```diff
@@ -57,7 +57,11 @@ using namespace llvm;
 #endif
 #define DEBUG_TYPE "lower-enzyme-intrinsic"
 
-void HandleAutoDiff(CallInst *CI, TargetLibraryInfo &TLI, AAResults &AA) {
+llvm::cl::opt<bool>
+    EnzymePostOpt("enzmye-postopt", cl::init(false), cl::Hidden,
+                  cl::desc("Run enzymepostprocessing optimizations"));
```
Comment (vchuravy) [inline enzyme/Enzyme/Enzyme.cpp:62]:
```suggestion
                  cl::desc("Run enzyme postprocessing optimizations"));
```

File: enzyme/Enzyme/Enzyme.cpp [hunk 2]
```diff
@@ -82,13 +86,17 @@ void HandleAutoDiff(CallInst *CI, TargetLibraryInfo &TLI, AAResults &AA) {
   unsigned truei = 0;
   IRBuilder<> Builder(CI);
 
+  bool AtomicAdd = llvm::Triple(CI->getParent()->getParent()->getParent()->getTargetTriple()).getArch() == Triple::nvptx ||
```
Comment (vchuravy) [inline enzyme/Enzyme/Enzyme.cpp:89]: Can't you look at the TLI instead?

File: enzyme/Enzyme/Enzyme.cpp [hunk 3]
```diff
@@ -128,6 +136,32 @@ void HandleAutoDiff(CallInst *CI, TargetLibraryInfo &TLI, AAResults &AA) {
...
+      if (MS == "enzyme_dup") {
+        ty = DIFFE_TYPE::DUP_ARG;
+        ++i;
+        res = CI->getArgOperand(i);
+      } else if (MS == "enzyme_dupnoneed") {
+        ty = DIFFE_TYPE::DUP_NONEED;
+        ++i;
+        res = CI->getArgOperand(i);
+      } else if (MS == "enzyme_out") {
+        llvm::errs() << "saw metadata for diffe_out\n";
```
Comment (vchuravy) [inline enzyme/Enzyme/Enzyme.cpp:154]: Make these `LLVM_DEBUG`?

File: enzyme/Enzyme/FunctionUtils.cpp [hunk 4]
```diff
@@ -335,9 +329,10 @@ Function *preprocessForClone(Function *F, AAResults &AA, TargetLibraryInfo &TLI,
 #if LLVM_VERSION_MAJOR > 10
         AM.registerPass([] { return PassInstrumentationAnalysis(); });
 #endif
-
+        #if LLVM_VERSION_MAJOR <= 7
```
Comment (vchuravy) [inline enzyme/Enzyme/FunctionUtils.cpp:332]:
```suggestion
#if LLVM_VERSION_MAJOR <= 7
```

File: enzyme/Enzyme/FunctionUtils.cpp [hunk 5]
```diff
@@ -335,9 +329,10 @@ Function *preprocessForClone(Function *F, AAResults &AA, TargetLibraryInfo &TLI,
...
         GVN().run(*NewF, AM);
         SROA().run(*NewF, AM);
+        #endif
```
Comment (vchuravy) [inline enzyme/Enzyme/FunctionUtils.cpp:335]:
```suggestion
#endif
```

File: enzyme/Enzyme/FunctionUtils.cpp [hunk 6]
```diff
@@ -692,18 +684,15 @@ void optimizeIntermediate(GradientUtils *gutils, bool topLevel, Function *F) {
 
   // TODO function attributes
   // PostOrderFunctionAttrsPass().run(*F, AM);
+  #if LLVM_VERSION_MAJOR <= 7
```
Comment (vchuravy) [inline enzyme/Enzyme/FunctionUtils.cpp:687]:
```suggestion
#if LLVM_VERSION_MAJOR <= 7
```

File: enzyme/Enzyme/FunctionUtils.cpp [hunk 7]
```diff
@@ -692,18 +684,15 @@ void optimizeIntermediate(GradientUtils *gutils, bool topLevel, Function *F) {
...
   GVN().run(*F, AM);
   SROA().run(*F, AM);
   EarlyCSEPass(/*memoryssa*/ true).run(*F, AM);
-#if LLVM_VERSION_MAJOR > 6
-  InstSimplifyPass().run(*F, AM);
-#endif
-  CorrelatedValuePropagationPass().run(*F, AM);
+  #endif
```
Comment (vchuravy) [inline enzyme/Enzyme/FunctionUtils.cpp:691]:
```suggestion
#endif
```

File: enzyme/Enzyme/GradientUtils.h [hunk 8]
```diff
@@ -1160,6 +1166,54 @@ assert(cast<PointerType>(ptr->getType())->getElementType() == dif->getType());
...
+                                            Intrinsic::nvvm_atomic_add_gen_f_sys, {dif->getType(), ptr->getType()}),
+                  {ptr, dif}));
+  } else
+  #endif
+  {
+    llvm::errs() << "unhandled atomic add: " << *ptr << " " << *dif << "\n";
+    llvm_unreachable("unhandled atomic add");
+  }
+  } else {
+    #if LLVM_VERSION_MAJOR >= 9
```
Comment (vchuravy) [inline enzyme/Enzyme/GradientUtils.h:1202]:
```suggestion
#if LLVM_VERSION_MAJOR >= 9
```

File: enzyme/Enzyme/GradientUtils.h [hunk 9]
```diff
@@ -1160,6 +1166,54 @@ assert(cast<PointerType>(ptr->getType())->getElementType() == dif->getType());
...
+    AtomicRMWInst::BinOp op = AtomicRMWInst::FAdd;
+    AtomicRMWInst* rmw = BuilderM.CreateAtomicRMW (op, ptr, dif, AtomicOrdering::Monotonic, SyncScope::System);
+    #if LLVM_VERSION_MAJOR >= 11
```
Comment (vchuravy) [inline enzyme/Enzyme/GradientUtils.h:1205]:
```suggestion
#if LLVM_VERSION_MAJOR >= 11
```

File: enzyme/Enzyme/Enzyme.cpp (hunk 3 above)
Comment (wsmoses) [reply enzyme/Enzyme/Enzyme.cpp:154]: removed

File: enzyme/Enzyme/Enzyme.cpp (hunk 2 above)
Comment (wsmoses) [reply enzyme/Enzyme/Enzyme.cpp:89]: TLI didn't seem to have a method to extract the triple (only via construction)

File: enzyme/Enzyme/FunctionUtils.cpp (hunk 4 above)
Comment (wsmoses) [reply enzyme/Enzyme/FunctionUtils.cpp:332]: Done

File: enzyme/Enzyme/FunctionUtils.cpp (hunk 5 above)
Comment (wsmoses) [reply enzyme/Enzyme/FunctionUtils.cpp:335]: Done

**Review by vchuravy — APPROVED** [review]

**Pull request merged**